Mathlib/Tactic/Change.lean

```diff
@@ -0,0 +1,39 @@
+/-
+Copyright (c) 2023 Damiano Testa. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Damiano Testa
+-/
+
+import Mathlib.Tactic.Basic
+/-!
+
+# Tactic `change? _`
+
+This tactic is used to replace the goal with a definitionally equal one.
+The `_` is a general term and is intended to be used to guide which unfolding you want
+Lean to perform.
+-/
+
+/--  `change? term?` unifies the optional `term?` with the current goal, defaulting to the
+goal, if `term?` is not present.
+It then prints in the infoView the unified term.
+This is useful to replace the main goal after a `dsimp`.
+```lean
+example : (fun x : Nat => x) 0 = 1 := by
+  change? 0 = _  -- `change 0 = 1`
+```
+-/
+syntax (name := change?) "change?" (ppSpace colGt term)? : tactic
+
+open Lean Meta Elab.Tactic Std.Tactic.TryThis in
+@[inherit_doc change?, tactic change?] def elabChange? : Tactic := fun stx => do
+  let sop := stx[1].getOptional?
+  let expr := ← match sop with
+    | none => getMainTarget
+    | some ex => do
+      let ex := ← elabTerm ex none
```

Comment:
Should we elaborate `ex` with the type of the main goal (instead of `none`), since we know it has to be defeq to the main goal? (There might be a reason not to, I'm not sure.)

Reply:
This seems like a good idea to me, but then I'd move the error about non-defeq to this step, right?  (I'll do it myself, but I won't get around to it for today.)

Reply:
By the way, I wonder if this change would fix what I consider a bug: with the running `fun` example, using `change? _ = 1` seems to ignore the hint and returns `fun ...`.

Reply:
I think It might actually be more informative to let `elabTermEnsuringType` report its own error! Then the user knows that the problem is specifically a type mismatch, not a term-level defeq problem.

Reply:
Ok, in any case, the explicit defeq test and the error then should be removed entirely, right?

Reply:
Hmm, I don't think so, if I understand correctly—we *do* still want to test that `ex` is defeq to the main target. `elabTermEnsuringType` just checks that these expressions have the same type.

Reply:
Also, if you do opt to click commit suggestion on the larger `elab_rules` suggestion, I'll update my suggestion here, as we should use `withRef` to report the type error on the term: e.g. `withRef sop <| elabTermEnsuringType ...`. Or if you'd like to make changes manually from your local repo, I'd recommend including `withRef` as appropriate (whatever's more convenient). :)

Reply:
I committed your elab_rules suggestion.  I plan to commit all the remaining as well.  I got into the habit of doing it while I am in front of a computer, to catch errors right away, but I'm happy to make exceptions!